include/boost/multiprecision/traits/max_digits10.hpp

```diff
@@ -0,0 +1,39 @@
+///////////////////////////////////////////////////////////////
+//  Copyright 2012 John Maddock. Distributed under the Boost
+//  Software License, Version 1.0. (See accompanying file
+//  LICENSE_1_0.txt or copy at https://www.boost.org/LICENSE_1_0.txt
+
+#ifndef BOOST_MATH_MAX_DIGITS10_HPP
+#define BOOST_MATH_MAX_DIGITS10_HPP
+
+namespace boost {
+namespace multiprecision {
+namespace detail {
+
+   template <unsigned digits>
+   struct calc_max_digits10
+   {
+#ifndef BOOST_NO_CXX11_CONSTEXPR
+      static constexpr unsigned constexpr_ceil(double d)
+      {
+         return static_cast<unsigned>(d) == d ? static_cast<unsigned>(d) : static_cast<unsigned>(d) + 1;
```

Comment:
about that `+ 1` at the end. I am [now doing](https://gitlab.com/cosurgi/minieigen-real/-/commit/54f3db9d6c0abe7360326ab20bcf6016deea4a48) some measurements for the value of how many [extra digits](https://gitlab.com/cosurgi/minieigen-real/-/blob/master/lib/high-precision/RealHPConfig.cpp#L14) are necessary to properly [convert back and forth](https://gitlab.com/cosurgi/minieigen-real/-/blob/master/lib/high-precision/RealIO.hpp#L34) and the results are quite surprising. The error consistently gets down. I thought that 3 or 4 is useful maximum. But so far, when calculating `sin(…)` in `long double` precision I got maximum number of incorrect decimal digits (calculated as log10 of error with respect to UMP (unit of least precision - the last binary digit)):

* for `1` error is 3.1 decimal digits
* for `2` error is 2.4 decimal digits
* for `3` error is 0.9 decimal digits
* for `4` error is 0.3 decimal digits
* for `5` error is 0 decimal digits

Reply:
I assume these are the additional digits required to get sin(x) correct to the last bit?

Which is somewhat different to what we have here: "how many decimal places do I need to print to be able to round-trip back to the binary" ?

Reply:
You are correct. In principle I wanted to test the mathematical functions - to make sure that [my mapping](https://gitlab.com/cosurgi/minieigen-real/-/blob/master/lib/high-precision/MathFunctions.hpp) finds the C++ code for implementation with correct precision. And that's how I found https://github.com/boostorg/math/issues/307

Afterwards I realized how much the error depends on [`extraStringDigits10`](https://gitlab.com/cosurgi/minieigen-real/-/blob/master/lib/high-precision/RealHPConfig.cpp#L14) and this forced me to make it configurable in runtime.

So yeah, it's a bit more than roundtripping. I am doing here (for all math functions):

xᵇᵃˢᵉ = random(-4π,4π) // other functions have a different range
vᵇᵃˢᵉ = sinᵇᵃˢᵉ(xᵇᵃˢᵉ)   // where ᵇᵃˢᵉ means that it's using the tested type, eg. `long double`
vᵐᵖᶠʳ = sinᵐᵖᶠʳ(binaryᵐᵖᶠʳ(string(xᵇᵃˢᵉ)))
errorᵐᵖᶠʳ = (binaryᵐᵖᶠʳ(string(vᵇᵃˢᵉ)) - vᵐᵖᶠʳ)/vᵐᵖᶠʳ
then I express error relative to ULP.

Yeah, I think that it would be enough to test only the round tripping part. However I am just curious how well the particular implementations of math functions perform, and I wanted to check that :)

> Which is somewhat different to what we have here: "how many decimal places do I need to print to be able to round-trip back to the binary" ?

I will add a round tripping only test to the list of my tests :)